tutorials/v7/forest/fst003_lhcbOpenData.C

```diff
@@ -0,0 +1,119 @@
+using RFieldBase = ROOT::Experimental::Detail::RFieldBase;
+using RInputForest = ROOT::Experimental::RInputForest;
+using ROutputForest = ROOT::Experimental::ROutputForest;
+
+constexpr char const* kTreeFileName = "http://root.cern.ch/files/LHCb/lhcb_B2HHH_MagnetUp.root";
+constexpr char const* kForestFileName = "fst003_lhcbOpenData.root";
+
+
+void Convert() {
+   std::unique_ptr<TFile> f(TFile::Open(kTreeFileName));
+   assert(f.is_valid());
+
+   // Get a unique pointer to an empty RForest model
+   auto model = RForestModel::Create();
+
+   // We create RForest fields based on the types found in the TTree
+   // This simple approach only works for trees with simple branches and only one leaf per branch
+   auto tree = f->Get<TTree>("DecayTree");
+   std::vector<TBranch*> branches;
+   for (TObject *obj : *(tree->GetListOfBranches())) {
+      auto b = static_cast<TBranch*>(obj);
+      // We assume every branch has a single leaf
+      TLeaf *l = static_cast<TLeaf*>(b->GetListOfLeaves()->First());
+
+      // Create a forest field with the same name and type than the tree branch
+      auto field = RFieldBase::Create(l->GetName(), l->GetTypeName());
+      std::cout << "Convert leaf " << l->GetName() << " [" << l->GetTypeName() << "]"
+                << " --> " << "field " << field->GetName() << " [" << field->GetType() << "]" << std::endl;
+
+      // Hand over ownership of the field to the forest model.  This will also create a memory location attached
+      // to the model's default entry, that will be used to place the data supposed to be written
+      model->AddField(std::unique_ptr<RFieldBase>(field));
+
+      // We connect the model's default entry's memory location for the new field to the branch, so that we can
+      // fill the forest with the data read from the TTree
+      void *fieldDataPtr = model->GetDefaultEntry()->GetValue(l->GetName()).GetRawPtr();
+      TBranch *branchRead = nullptr;
+      tree->SetBranchAddress(b->GetName(), fieldDataPtr);
+      branches.push_back(branchRead);
+   }
+
+   // The new forest takes ownership of the model
+   auto forest = ROutputForest::Recreate(std::move(model), "DecayTree", kForestFileName);
+
+   auto nEntries = tree->GetEntries();
+   for (decltype(nEntries) i = 0; i < nEntries; ++i) {
+      tree->GetEntry(i);
+      forest->Fill();
+
+      if (i && i % 100000 == 0)
+         std::cout << "Wrote " << i << " entries" << std::endl;
+   }
+}
+
+
+void fst003_lhcbOpenData()
+{
+   if (gSystem->AccessPathName(kForestFileName))
+      Convert();
+
+   // Create histogram of the flight distance
+
+   // We open the forest without specifiying an explicit model first, but instead use a view on the field we are
+   // interested in
+   auto forest = RInputForest::Open("DecayTree", kForestFileName);
+
+   // The view wraps a read-only double value and accesses directly the forest's data buffers
+   auto viewFlightDistance = forest->GetView<double>("B_FlightDistance");
```

Comment:
What is the semantic difference between setting a read-model and using a view?  (i.e. do we need both)?

Reply:
With `LoadEntry()`, data is copied into user-controlled memory referenced in an `REntry`.  The view provides a channel directly into the page buffers, at least for simple types.

Defining a read model and loading entries should be more convenient, for instance one can write to the memory location referenced by an entry.  The view is potentially faster but more difficult to use.

Reply:
Fair enough.